### PR TITLE
Design flaw BoundingBox class

### PR DESCRIPTION
**Pull request by jessearens**
This pull request goes combined with the code metrics assignment. For extra information, see Assignment 4 (which can be found our Google Drive for now, and will be uploaded to this repository when all parts of this assignment are finished).
inCode showed us a design flaw in the BoundingBox class, namely the schizophrenic class.
The biggest problem with the BoundingBox class according to this design flaw and inCode's view on it, is the fact that the intersects(CollisionArea other) method did not belong to the same cluster as all others. It did not use common fields with other methods and also was not called by the same methods as the other methods in BoundingBox. Therefore, intersects(CollisionArea other) was kind of the schizophrenic second personality of the BoundingBox.

After taking a look into this method, I found out that it's only reason to still be in the BoundingBox class was because of the interface that specified it should be there. I have removed the intersects method from both BoundingBox and ICollisionArea. To be able to do so, I have also had to make PlayerFish use a CollisionMask instead of a random ICollisionArea, which is what we should want anyway. 

Please note that quite some tests still depended on the intersects method of a BoundingBox, even though the real in-game implementation did not use this method anymore. I refactored most of them to use the intersects method of a CollisionMask. TestEnemyFish however was still written as if though EnemyFish used a BoundingBox, even though it does not use that anymore. Therefore it was enormously outdated and I have deleted TestEnemyFish for now. A new test using CollisionMasks still has to be written.

<organisation

### CODE REVIEW

File: src/main/java/com/github/fishio/Entity.java [hunk 1]
```diff
@@ -80,12 +80,12 @@ public void kill() {
 	
 	
 	@Override
-	public ICollisionArea getBoundingArea() {
-		return boundingArea;
+	public CollisionMask getBoundingArea() {
+		return (CollisionMask) boundingArea;
```
Comment (LMiljak) [inline src/main/java/com/github/fishio/Entity.java]: Unnecessary cast. Just change the type of the boundingArea attribute from ICollisionArea to CollisionMask.

<logic

**Comment by LMiljak** [comment]
This all looks good and certainly fixes the BoundingBox being a schizophrenic class. However, the problem with BoundingBox is simply that it isn't used ANYWHERE anymore, except for the tests. Wouldn't removing it simply solve everything? I know it sounds a little bit strange, just deleting everything that causes a problem, but if it's useless, why keep it in our project? That's just my suggestion. 

We basicly don't want any other ICollisionArea except for CollisionMask, because we can only check for intersections between two of the same types (it's hard to check for intersections between a CollisionMask object and a BoundingBox object, that's why we changed everything from BoundingBox to CollisionMask in the first place). And because we only want to have the CollisionMask as the only ICollisionArea, we might as well remove the interface! This is just my suggestion, I know it sounds really too simple, but consider it. We don't need and don't want BoundingBoxes anymore, because only CollisionMasks are supported now.

<organisation, logic

**Comment by LMiljak** [comment]
Actually, now that I gave it a second thought. The BoundingBox class can still be useful in the future, but certainly NOT for collisions, we still want only CollisionMasks for that. The reason for this is that the BoundingBox has a position, which can be used if we want to have "entities" that don't have collisions (for example making a submarine render in the background). However, what I still find important then is that ICollisionArea should be renamed to IBoundingArea, because we don't want BoundingBoxes to do something with collisions.
In this case, you would barely have to change anything apart from the name of ICollisionArea.

<organisation, logic, naming

File: src/main/java/com/github/fishio/ICollisionArea.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ default boolean isReversed() {
 	 * @return true if this bounding box collides with the given BoundingArea,
 	 *         false if not.
 	 */
-	boolean intersects(ICollisionArea other);
```
Comment (Taeir) [inline src/main/java/com/github/fishio/ICollisionArea.java:78]: Don't forget to remove the javadoc as well.

<documentation

**Comment by Taeir** [comment]
There is one problem with using collision masks: they are a lot harder to create. In order to create them, all tests would have to extend SlimGuiTest and use images of the correct size.

This would make testing simple things much harder.

I agree that the current situation is not preferrable, so perhaps we should also refactor collisionmask to support simple masks. Perhaps a constructor where all data is set to true (aka the CollisionMask is a box).

TL DR: This pull request identifies a much larger problem which should be addressed soon. I will create an issue for this.

<organisation

**Comment by jessearens** [comment]
Fixed the comments. I agree that BoundingBox should be obsolete, but for now it's okay to keep it to this refactoring to at least make the class non-schizophrenic.

<organisation

**Comment by LMiljak** [comment]
Very well then. I'm merging it. This should be resolved next sprint.

<organisation
